### ISSUE TITLE
Feature request: complementary bone in Outfit Studio

### ISSUE THREAD

**Issue by sts1skj** [ousnius/BodySlide-and-Outfit-Studio#225]
**What is it that you want to do that isn't currently possible?**
Edit bone weights effectively.

**Request the feature**
Feature 1:  I suggest adding a "complementary bone" selection option when editing bone weights in Outfit Studio.  Suppose I select bone A to edit its weights.  I would like to be able to designate a "complementary bone" B.  When the weight for bone A is increased by x for a vertex V, B's weight for V would be decreased by x.  The weights of A and B for the vertex would be restricted to [0,1], so, for example, if A's weight at V started out at 0 and B's weight started out at .5, A's weight could not be increased past .5 (at which point B's weight would be 0).

Use case 1a: I want to create a weight set for the bone "NPC L Butt" by taking weight away from "NPC Pelvis [Pelv]".  I want to have the changes limited to the existing weight, so I can't accidentally "paint outside the lines".  I want the pelvis weight automatically reduced by the amount I add to the butt.

Use case 1b: When copying the bone vertex weights of "NPC L Butt" from one model to another, I want the changes to the weight of this bone on the target subtracted from the target's pelvis weights so that I can preserve the existing weights on the target as well as possible and only modify "NPC L Butt" and "NPC Pelvis [Pelv]".

Feature 2 (contingent on feature 1): I suggest adding a bone-weight-scaling function.  With a bone A and a complementary bone B selected, the user would enter a scaling factor x.  All of the vertex weights for A would be multiplied by x, with the difference added or subtracted from B's vertex weights.

Use case 2: Suppose I've used feature 1 (use case 1b) to copy the "NPC L Butt" weights from my body model to a clothing model.  I want the butt to move 80% as much when clothed as when naked, so I'd like to scale the "NPC L Butt" weights by multiplying each vertex's weight by .8, with the lost weight transferred to "NPC Pelvis [Pelv]".

Feature 3 (contingent on feature 1): I suggest adding bone selections for "mirror bone" and "complementary mirror bone".  Any change made to a vertex's bone weight would also be made to the mirror bone's weight for the mirror vertex, with the complementary and complementary mirror bone weights updated appropriately.  The complementary bone and the complementary mirror bone may or may not be the same bone.

Use case 3 (continuing use case 1a): When I create the vertex weights for "NPC L Butt", I want weights created for "NPC R Butt" too so that the two sets of bone weights are perfect mirrors of each other.  I want both sets of bone weights to be subtracted from their corresponding complements (which, in this case, are the same: "NPC Pelvis [Pelv]").

Feature 4 (alternative to feature 3, contingent on feature 1): I suggest adding a mirror-copy-bone-weight function that can copy the weights for one bone to another within the same model, with the vertices affected mirrored.

Use case 4 (continuing use case 1a): After creating the new vertex weights for "NPC L Butt", I want to make "NPC R Butt" have exactly the same weights mirrored across the body.  The changes to the target bone's weights should also affect the complementary bone's weights, in this case "NPC Pelvis [Pelv]".

Thoughts: In the existing Outfit Studio, any vertex's weight for a bone can be set to any value between 0 and 1.  Since this usually results in a sum of the weights for a vertex that is not 1, Outfit Studio normalizes the weights when the file is saved.  This often has unexpected results, with bone weights that were previously smooth turning jagged during the save process.  In addition, it's too easy to "paint outside the lines", resulting in the wrong vertices moving when a bone moves.  I've spent quite a bit of time trying to create satisfying butt and thigh weights for the body model in Skyrim, and the results of my work are surprisingly awful.  (The existing body shape mods are not much better, in my opinion.)  I'm not certain what the best solution is, but I think it's likely that the features I've suggested in this message would help a lot.

Being able to see the effects of any weight changes immediately would be great too, of course.  I'll write a second feature request for that.

**Comment by ousnius** [comment]
Your feature requests things can be compiled into three things:
1. Normalizing weights actively while painting
2. Restricting painting to already weighted vertices (don't edit 0.0 weight)
3. Mirroring weights (select source and destination)

I don't think a selection of "complementary bone" is really needed, as what you really want is a normalization of the weights as you paint instead of on export. Selecting one bone as a complementary bone isn't enough either, as not just the pelvis bone needs to have its weights reduced as you add to the butt bone, but also all other bones affecting the same vertices as well.

**Comment by sts1skj** [comment]
I'm hardly an expert on editing weights, but my instincts say that it's important to not change the weights for bones other than the complementary bone when changing the weight of the selected bone.

Here's an example: in the Skyrim body model "CBBE Model", the left buttock is primarily covered by two bone weights: "NPC Pelvis [Pelv]" and "NPC L Thigh [LThg]".  The added bone for the buttock would be "NPC L Butt".  This new bone is constrained to stay close to "NPC Pelvis [Pelv]"; in the default case, it is fixed relative to that bone.  It has no constraint relative to the thigh.  If, when creating the weights for "NPC L Butt", the weight for "NPC L Thigh [LThg]" is reduced, then the weighting of vertices towards the pelvis increases relative to the weighting towards the thighs.  The result is that the butt deforms unnaturally as the thigh moves relative to the pelvis.

My gut says that, in general, when adding weights for a new bone, the complementary bone should be the bone the new bone is attached to, in a motion-constraint sense.

Here's another example (that I actually ran into a few days ago):  The left knee joint in "CBBE Model" is covered by a blend of "NPC L Thigh [LThg]" and "NPC L Calf [LClf]".  If, in creating "NPC L Front Thigh", I take any weight away from "NPC L Calf [LClf]" (even indirectly through the automatic normalization on save), this screws up the appearance of the knee.  A surprisingly small change to the relative weighting between calf and thigh causes the knee to deform horribly when the character squats.

**Comment by ousnius** [comment]
Yeah, of course only weights for bones in the same hierarchy as the edited bone should be part of the live normalization. The weights of a vertex together should then only be normalized to a sum of 1.0 on save (where the visuals wouldn't change anymore after live normalization). This would have to be tested with some NIFs in-game though to confirm.

**Comment by sts1skj** [comment]
@ousnius: I could start looking at implementing this.  Would you like me to?

We seem to disagree on the best approach.  I could write a proposal, either brief or detailed.

**Comment by ousnius** [comment]
@sts1skj: The first step to this is to add interactive weight normalization on mouse up (propagating to the "AnimInfo" object and the displayed mesh's vertex colors) the same way e.g. Max and Maya do it.

Making weights stronger or weaker should substract the changed factor off of all other bones that affect the same vertices. This should happen on mouse up after changing weights with the weight brush and after copying bone weights.

As a first step I wouldn't do it every frame while painting weights as that causes more user confusion.
Then we can talk about mirroring, locking, complementary bones or whatever you want to call such features.

**Comment by ousnius** [comment]
Blender auto normalize for reference:
https://github.com/blender/blender/blob/master/source/blender/editors/sculpt_paint/paint_vertex.c#L830

**Comment by sts1skj** [comment]
I've been studying the Outfit Studio source code, trying to figure out how to implement continuous weight normalization, and I've found something that I think ought to be done first: TweakStroke cleanup.

The TweakStroke class, as currently written in Outfit Studio, has two purposes: storing information about a change to the project so that that change can be applied, undone, or redone; and performing the work of a brush stroke.  It seems to me that TweakStroke ought to be split into two classes, one for each purpose.  Partly this is because it seems cleaner to me from an object-oriented programming point of view; and partly it's because not all changes that can be undone are brush strokes.

I could make these changes to TweakStroke, if you like.

So I could:
-try to implement continuous weight normalization without changing how TweakStroke works.
-clean up TweakStroke.
-work on something else.
-wait for you to clean up TweakStroke before doing anything.

What shall I do?

**Comment by ousnius** [comment]
You're free to try to clean up TweakStroke first, but try not to change any behavior.

**Comment by sts1skj** [comment]
At this point, features 1 and 3 in my original post are complete.

I no longer feel that feature 2 (weight scaling) would be useful.  I've experimented quite a bit, and I now feel that the amount of motion is best adjusted using physics simulation parameters, not bone weight scaling.  (It seems to work best if the bone weights for a breast, butt, or thigh are scaled so that the peak weight is close to 1.)  Unfortunately, with the current state of HDT-SMP for Skyrim, this means having a distinct physics file for each group of outfits with different types of motion, but that's a weakness in HDT-SMP, not Outfit Studio.

Feature 4 (copying weights to a mirror) might be somewhat useful, but I've found in practice that feature 2 (x-mirror bone) is sufficient.  I'm skeptical that a mirror-weight-copy function could produce good results unless the mesh were perfectly symmetrical; I've found vertex bone weighting to be very sensitive, with tiny errors producing visible pose-related rough spots.  I could implement feature 4 if you want, but I'd rather focus on other features right now.

So I'm okay with this feature request being closed.
